invoices/invoiceregistry.go

```diff
@@ -938,6 +938,7 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 		customRecords:        payload.CustomRecords(),
 		mpp:                  payload.MultiPath(),
 		amp:                  payload.AMPRecord(),
+		metadata:             payload.Metadata(),
```

Comment:
What is missing here is that we are assuming that an invoice already exists. The goal of the new metadata field is to enable stateless invoices. In that case, there isn't an invoice yet. It is a spontaneous payment, but the difference with keysend/amp is that the receiver is able to re-derive the preimage. The preimage can be used as proof of payment.

I am wondering what the options are to implement this. Will it be yet another payment type? Or is this the point where it becomes necessary to allow invoice registry plugins so that users can implement the logic themselves?

Reply:
LDK approach with this: https://github.com/rust-bitcoin/rust-lightning/issues/1171

Reply:
> I am wondering what the options are to implement this. Will it be yet another payment type

I think we can just use the _existing_ keysend flow here, which ends up inserting the the database, or using the existing AMP pattern to re-insert a new sub-invoice with a main tracking identifier. IIUC, with the proposal you'd still write the invoice to disk, but the "stateless" part is what lets you not have to remember the invoice until its paid?

Reply:
Alternatively, the existing HTLC interceptor logic can handle this as well assuming you have the invoice logic already living outside of lnd.

Reply:
I think "stateless" can be implemented in multiple ways, but doing a JIT insert similar to keysend seems definitely one of the options.